### ISSUE TITLE
nxos_switchport not configuring trunk ports properly

### ISSUE THREAD

**Issue by zoltan** [ansible/ansible-modules-core#4082]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

nxos_switchport
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

NX-OS version 7.0(3)I2(2a)
```
##### CONFIGURATION

Using defaults:

```
ansible@maas:~/ansible$ cat /etc/ansible/ansible.cfg  | grep -v ^# | grep -v ^$
[defaults]
[privilege_escalation]
[paramiko_connection]
[ssh_connection]
[accelerate]
[selinux]
[colors]
ansible@maas:~/ansible$
```
##### OS / ENVIRONMENT

Ubuntu 14.04, ansible installed from ppa, but the issue is not platform specific.
##### SUMMARY

Trying to get a switchport to be configured for trunk however 
- the "switchport access" configuration is not removed even with the state=absent action
- the actual trunk vlan list is not applied to the interface
##### STEPS TO REPRODUCE

```
Playbook:
nsible@maas:~/ansible$ cat cisco.yml

---

- name: vlan provisioning
  hosts: cisco
  connection: local

  tasks:
    - nxos_switchport: interface=port-channel3123 mode=access state=absent host={{ inventory_hostname }} username="user" password="password"
    - nxos_switchport: interface=port-channel3123 mode=trunk native_vlan=113 trunk_vlans=10-11 host={{ inventory_hostname }} username="user" password="password"
ansible@maas:~/ansible$
```
##### EXPECTED RESULTS

```
interface port-channel3123
  description c05n2
  switchport mode trunk
  switchport trunk native vlan 113
  switchport trunk allowed vlan 10-11
  spanning-tree port type edge
  mtu 9216
  vpc 3154
```
##### ACTUAL RESULTS

Before running the playbook:

```
interface port-channel3123
  description c05n2
  switchport access vlan 113
  spanning-tree port type edge
  mtu 9216
  vpc 3123
```

After running the playbook:

```
interface port-channel3123
  description c05n2
  switchport mode trunk
  switchport access vlan 113
  switchport trunk native vlan 113
  spanning-tree port type edge
  mtu 9216
  vpc 3123
```

Output from the playbook:

```
ansible@maas:~/ansible$ ansible-playbook -i hosts cisco.yml -vvvvvvvvvvv
Using /etc/ansible/ansible.cfg as config file
Loaded callback default of type stdout, v2.0

PLAYBOOK: cisco.yml ************************************************************
1 plays in cisco.yml

PLAY [vlan provisioning] *******************************************************

TASK [setup] *******************************************************************
<ip> ESTABLISH LOCAL CONNECTION FOR USER: ansible
<ip> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1467214096.91-180213448017670 `" && echo ansible-tmp-1467214096.91-180213448017670="` echo $HOME/.ansible/tmp/ansible-tmp-1467214096.91-180213448017670 `" ) && sleep 0'
<ip> PUT /tmp/tmpMxf1kk TO /home/ansible/.ansible/tmp/ansible-tmp-1467214096.91-180213448017670/setup
<ip> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/ansible/.ansible/tmp/ansible-tmp-1467214096.91-180213448017670/setup; rm -rf "/home/ansible/.ansible/tmp/ansible-tmp-1467214096.91-180213448017670/" > /dev/null 2>&1 && sleep 0'
ok: [cisco1]

TASK [nxos_switchport] *********************************************************
task path: /home/ansible/ansible/cisco.yml:8
<ip> ESTABLISH LOCAL CONNECTION FOR USER: ansible
<ip> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1467214097.53-61932177549416 `" && echo ansible-tmp-1467214097.53-61932177549416="` echo $HOME/.ansible/tmp/ansible-tmp-1467214097.53-61932177549416 `" ) && sleep 0'
<ip> PUT /tmp/tmpuViqHO TO /home/ansible/.ansible/tmp/ansible-tmp-1467214097.53-61932177549416/nxos_switchport
<ip> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/ansible/.ansible/tmp/ansible-tmp-1467214097.53-61932177549416/nxos_switchport; rm -rf "/home/ansible/.ansible/tmp/ansible-tmp-1467214097.53-61932177549416/" > /dev/null 2>&1 && sleep 0'
ok: [cisco1] => {"changed": false, "end_state": {"access_vlan": "113", "access_vlan_name": "test", "interface": "port-channel3123", "mode": "access", "native_vlan": "1", "native_vlan_name": "default", "switchport": "Enabled", "trunk_vlans": "1-4094"}, "existing": {"access_vlan": "113", "access_vlan_name": "test", "interface": "port-channel3123", "mode": "access", "native_vlan": "1", "native_vlan_name": "default", "switchport": "Enabled", "trunk_vlans": "1-4094"}, "invocation": {"module_args": {"access_vlan": null, "host": "cisco1", "interface": "port-channel3123", "mode": "access", "native_vlan": null, "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "port": null, "provider": null, "ssh_keyfile": null, "state": "absent", "transport": "cli", "trunk_vlans": null, "use_ssl": false, "username": "ansible", "validate_certs": true}, "module_name": "nxos_switchport"}, "proposed": {"interface": "port-channel3123", "mode": "access"}, "state": "absent", "updates": []}

TASK [nxos_switchport] *********************************************************
task path: /home/ansible/ansible/cisco.yml:9
<ip> ESTABLISH LOCAL CONNECTION FOR USER: ansible
<ip> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1467214126.46-150234195272525 `" && echo ansible-tmp-1467214126.46-150234195272525="` echo $HOME/.ansible/tmp/ansible-tmp-1467214126.46-150234195272525 `" ) && sleep 0'
<ip> PUT /tmp/tmpcUsQx4 TO /home/ansible/.ansible/tmp/ansible-tmp-1467214126.46-150234195272525/nxos_switchport
<ip> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/ansible/.ansible/tmp/ansible-tmp-1467214126.46-150234195272525/nxos_switchport; rm -rf "/home/ansible/.ansible/tmp/ansible-tmp-1467214126.46-150234195272525/" > /dev/null 2>&1 && sleep 0'
changed: [cisco1] => {"changed": true, "end_state": {"access_vlan": "113", "access_vlan_name": "test", "interface": "port-channel3123", "mode": "trunk", "native_vlan": "113", "native_vlan_name": "test", "switchport": "Enabled", "trunk_vlans": "1-4094"}, "existing": {"access_vlan": "113", "access_vlan_name": "test", "interface": "port-channel3123", "mode": "access", "native_vlan": "1", "native_vlan_name": "default", "switchport": "Enabled", "trunk_vlans": "1-4094"}, "invocation": {"module_args": {"access_vlan": null, "host": "cisco1", "interface": "port-channel3123", "mode": "trunk", "native_vlan": "113", "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "port": null, "provider": null, "ssh_keyfile": null, "state": "present", "transport": "cli", "trunk_vlans": "10-11", "use_ssl": false, "username": "ansible", "validate_certs": true}, "module_name": "nxos_switchport"}, "proposed": {"interface": "port-channel3123", "mode": "trunk", "native_vlan": "113", "trunk_vlans": "10-11"}, "state": "present", "updates": ["interface port-channel3123", "switchport mode trunk", "switchport trunk native vlan 113"]}

PLAY RECAP *********************************************************************
cisco1  : ok=3    changed=1    unreachable=0    failed=0

ansible@maas:~/ansible$
```

**Comment by GGabriele** [comment]
Actually, these are not bugs, that's just how the module is expected to work.

_The "switchport access" configuration is not removed even with the state=absent action_
There you have an access vlan, so you have to specify the `access_vlan` you want to remove.

_The actual trunk vlan list is not applied to the interface_
From the DOCSTRING:

```
When working with trunks VLANs the keywords add/remove are always sent in the `switchport trunk allowed vlan` command. Use verbose mode to see commands sent.
```

This means that the command sent is `switchport trunk allowed vlan add/remove #`, not `switchport trunk allowed vlan #`.
Basically here you can only add/remove allowed vlans.

I think the following will work for you:

##### Tasks

```
  - name: REMOVE ACCESS VLAN
    nxos_switchport: interface=eth1/5 mode=access access_vlan=123 state=absent host={{ inventory_hostname }} username="user" password="password"

  - name: CONFIGURE TRUNK
    nxos_switchport: interface=eth1/5 mode=trunk native_vlan=113 host={{ inventory_hostname }} username="user" password="password"

  - name: REMOVE ALLOWED VLANS 1-9
    nxos_switchport: interface=eth1/5 mode=trunk trunk_vlans=1-9 state=absent host={{ inventory_hostname }} username="user" password="password"

  - name: REMOVE ALLOWED VLANS 12-4094
    nxos_switchport: interface=eth1/5 mode=trunk trunk_vlans=12-4094 state=absent host={{ inventory_hostname }} username="user" password="password"
```

##### Before the run

```
interface Ethernet1/5
  switchport access vlan 123
  spanning-tree port type edge
```

##### After the run

```
interface Ethernet1/5
  switchport mode trunk
  switchport trunk native vlan 113
  switchport trunk allowed vlan 10-11
  spanning-tree port type edge
```

**Comment by jedelman8** [comment]
What @GGabriele said is all accurate.  You could also do state=absent trunk_vlans=2=4096 as you bootstrap all interfaces, then do what you wanted to originally.  Understandably, that may not be what you desire, but I think longer term one approach would be to have a flag that offers "allowed" or "add" - need to think this through and test.

Regarding the extra command in the 'show' output, that is purely cosmetic.  You can always check operational status with 'show interface switchport' commands.  This would be a lower priority than the first item, but I do understand too not wanting to see that extra command in there.

**Comment by ansibot** [comment]
@privateip, @GGabriele, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by GGabriele** [comment]
This can be closed

**Comment by zoltan** [comment]
Thanks for the explanations and sorry for the late reply. I think the current behaviour is not THAT intuitive, and having those flags you mentioned @jedelman8 would be great.

Thanks again!

**Comment by gundalow** [comment]
Is this something we can make clearer by improving the documentation?

**Comment by jedelman8** [comment]
@gundalow Yes, of course.  Will try sending a PR in tonight for code and doc updates.
